### PR TITLE
test-infra: add postgres discovery fallback when testcontainers exec is blocked

### LINKED ISSUE #360: test-infra: add postgres discovery fallback when testcontainers exec is blocked
## Scope spillover from omniscient/markethawk#288

The dark factory noticed this test-infrastructure gap while implementing issue omniscient/markethawk#288 but did not address it inline (scope enforcement — it's pre-existing infra, not part of the scanner pipeline spec).

**File/area:** backend/tests/conftest.py
**Defect:** In Docker environments where the socket proxy blocks exec calls (EXEC:0), testcontainers cannot run health checks and the transaction-rollback db fixture cannot spin up a test PostgreSQL. The _testcontainers_url() context manager has no fallback to a running postgres discovered from Docker API or DNS.

The omniscient/markethawk#288 implementation included a _probe_running_postgres() workaround inline. The proper fix is to make this discovery mechanism a first-class part of the test infrastructure so all tests using the db fixture work in the factory environment without per-implementation workarounds.

---
*Automatically triaged by MarketHawk Dark Factory scope enforcement.*

### PR DESCRIPTION
**Pull request by omniscient**
Closes #360

## Summary
Automated implementation for issue #360.

## Preview
_No preview environment — this change does not affect the running app ('All changes are test-only: refactoring postgres discovery utilities from conftest.py into a dedicated backend/tests/utils module with added test coverage. No application code, migrations, dependencies, or running services touched.')._

## Commands
```bash
# Iterate after feedback
docker compose --profile factory run --rm dark-factory "Continue issue #360"

# Tear down preview when done
docker compose --profile factory run --rm dark-factory "Close issue #360"
```

---
*Generated by MarketHawk Dark Factory*

### CODE REVIEW

File: backend/tests/utils/pg_discovery.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    explicit = os.environ.get("TEST_DATABASE_URL")
+    if explicit:
+        return explicit
+
+    # Collect candidate IPs from running postgres containers via Docker API.
+    candidate_ips: list[str] = []
+    docker_host = os.environ.get("DOCKER_HOST", "")
+    if docker_host.startswith("tcp://"):
+        try:
+            r = requests.get(
```
Comment (omniscient) [inline backend/tests/utils/pg_discovery.py:32]: **edge-case** · **medium** — `requests.get(...).json()` has no `raise_for_status()` and assumes the body is a list of containers; a proxy returning a non-200 JSON object (e.g. `{"message": "..."}`) makes the `for c in r.json()` iterate dict keys (strings) and `c.get(...)` raises `AttributeError`, caught by the bare `except` — so an authorization/error response is indistinguishable from "no containers". Add `r.raise_for_status()` and guard `isinstance(r.json(), list)`.

File: backend/tests/utils/pg_discovery.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                f"http://{docker_host[6:]}/containers/json",
+                timeout=3,
+            )
+            for c in r.json():
+                if "postgres" not in c.get("Image", "").lower():
+                    continue
```
Comment (omniscient) [inline backend/tests/utils/pg_discovery.py:38]: **maintainability** · **low** — The outer `except Exception: pass` around Docker API discovery silently drops every failure; add a `logging.debug(...)` so factory debugging isn't blind.

File: backend/tests/utils/pg_discovery.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    for hostname in ["postgres", "stockscanner-db", "localhost"]:
+        candidate_ips.append(hostname)
+
+    common_creds = [
+        ("postgres", "postgres", "postgres"),
+        ("postgres", "postgres", "stockscanner"),
+        ("onecli", "onecli", "onecli"),
+    ]
+    for ip in candidate_ips:
+        for user, pw, db in common_creds:
```
Comment (omniscient) [inline backend/tests/utils/pg_discovery.py:57]: **security** · **low** — Hardcoded credential triples are fine in test-only discovery code, but add a comment clarifying these are well-known dev defaults so they aren't mistaken for leaked secrets in future audits.

File: backend/tests/utils/pg_discovery.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            try:
+                psycopg2.connect(
+                    host=ip,
+                    port=5432,
```
Comment (omniscient) [inline backend/tests/utils/pg_discovery.py:61]: **maintainability** · **low** — The per-credential `except Exception: pass` is expected (probing), but emit a single debug log when all candidates fail so a CI run that finds no Postgres leaves a breadcrumb.

**Review by omniscient — COMMENTED** [review]
## 🏭 Dark Factory Code Review

**0** blocking · **4** advisory finding(s).

**Comment by omniscient** [comment]
## Dark Factory Run — Continued iteration

**PR:** #414

**Branch:** `feat/issue-360-test-infra--add-postgres-discovery-fallb`

### Changes

# Implementation Summary — Issue #360 (continue)

## Blocking Finding Fixed

Addressed: `_testcontainers_url` in `conftest.py` called `probe_running_postgres()` unconditionally. If the probe matched the dev postgres (`stockscanner-db`), `db_engine`'s `drop_all(bind=engine)` teardown would destroy its tables.

**Fix:** Gated the probe call behind an explicit opt-in flag: `POSTGRES_DISCOVERY_ENABLED=true`. Normal CI never sets this flag, so it always falls through to an ephemeral testcontainers DB. The factory environment sets `POSTGRES_DISCOVERY_ENABLED=true` explicitly to activate the probe.

## Files Modified

- `backend/tests/conftest.py` — added `POSTGRES_DISCOVERY_ENABLED=true` guard around the `probe_running_postgres()` call in `_testcontainers_url()`

## Tests

All 5 unit tests in `tests/utils/test_pg_discovery.py` pass.

## Migrations

None — test-only change.

### Commands
```bash
# Tear down when done
docker compose --profile factory run --rm dark-factory "Close issue #360"
```

---
*Posted by MarketHawk Dark Factory*
